### DISCUSSION TITLE
Time-aware meters

### DISCUSSION THREAD

**Discussion by BFergerson** [apache/skywalking#8121] [Ideas]
Meters appear to be only as granular as the `Config.Meter.REPORT_INTERVAL`. This means if you create a meter, like a counter, and trigger it once per second, what actually gets saved is a count of `90` per minute time bucket. Since I'm triggering something that occurs once per second, I'd expect the count per minute to be `60`.

I'm curious if anyone has any alternative solutions.
Changing `Config.Meter.REPORT_INTERVAL` to `1` is likely to work, but would be obviously inefficient.

**Comment by wu-sheng** [comment]
I think how to save counter depending on MAL engine. For counter, in best practice, it is recommended to use rate function to calculate the increasement, and then use sum/avg to get final result.

**Reply by BFergerson** [reply]
I believe what you're saying is to use multiple meters to gather the stats I'm looking for. That I should dedicate one meter to tracking the frequency and another for the total. Is that correct? If so, wouldn't that be less efficient than allowing meters to record and report multiple time buckets at a time?

**Reply by wu-sheng** [reply]
I mean one meter, but use rate to calculate rather than using raw metric data.

**Reply by BFergerson** [reply]
I don't fully understand why it works, but I can confirm using `rate` mode instead of `increment` mode works. Thanks

**Reply by wu-sheng** [reply]
It means, `rate` = `Tn - T1`. Such as in one minute you have `Tn1 - Tn10`, the default downsampling(avg) is doing avg from `Tn1- T1` to `Tn10 - T1`. No matter `Tn1` to `Tn10` are same or changed from time to time, the final result provides you expected data.
